File: 0299_bulls_and_cows/get_hint.py
from collections import defaultdict


def get_hint(secret: str, guess: str) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    h = defaultdict(int)
    bulls = cows = 0
    for idx, s in enumerate(secret):
        g = guess[idx]
        if s == g:
            bulls += 1
        else:
            cows += int(h[s] < 0) + int(h[g] > 0)
            h[s] += 1
            h[g] -= 1
    return "{}A{}B".format(bulls, cows)

File: 0299_bulls_and_cows/get_hint.java
import java.util.HashMap;

class Solution {
    // Time: O(n)
    // Space: O(n)
    public String getHint(String secret, String guess) {
        HashMap<Character, Integer> hm = new HashMap<Character, Integer>();
        int bulls = 0, cows = 0;
        for (int i = 0; i < secret.length(); i++) {
            char s = secret.charAt(i);
            char g = guess.charAt(i);
            if (s == g) {
                bulls++;
            } else {
                if (hm.getOrDefault(s, 0) < 0) {
                    cows++;
                }
                if (hm.getOrDefault(g, 0) > 0) {
                    cows++;
                }
                hm.put(s, hm.getOrDefault(s, 0) + 1);
                hm.put(g, hm.getOrDefault(g, 0) - 1);
                System.out.println(hm);
            }
        }
        return Integer.toString(bulls) + "A" + Integer.toString(cows) + "B";
    }
}